### PR TITLE
Allow for SFTP instead of FTP

### PR DESCRIPTION
**Pull request by zfrenchee**
I moved this from my master branch into an actual branch and the previous (#83) PR disappeared.

### CODE REVIEW

**Comment by zfrenchee** [comment]
As mentioned in #83, I think this is just about everything needed to make it go. How do I test this? I've made these changes on my personal galaxy and it works, but I'd like to test it here too, if possible.

File: tasks/proftpd.yml [hunk 1]
```diff
@@ -15,3 +15,6 @@
     - LoadModule mod_sql.c
     - LoadModule mod_sql_passwd.c
     - LoadModule mod_sql_postgres.c
+    - LoadModule mod_sftp.c
+    - LoadModule mod_sftp_pam.c
```
Comment (mvdbeek) [inline tasks/proftpd.yml]: Could you make a separate task for the sftp modules, that fires when `proftpd_use_sftp ` is true?

File: templates/proftpd.conf.j2 [hunk 2]
```diff
@@ -17,6 +17,18 @@ Group                           nogroup
 # Passive port range for the firewall
 PassivePorts                    {{ proftpd_passive_port_low }} {{ proftpd_passive_port_high }}
 
+{% if proftpd_use_sftp_instead %}
+SFTPEngine                      on
+
+# Configure both the RSA and DSA host keys, using the same host key files that OpenSSH uses.
+SFTPHostKey                     /etc/proftpd/ssh_host_keys/rsa
+SFTPHostKey                     /etc/proftpd/ssh_host_keys/dsa
+SFTPAuthMethods                 publickey keyboard-interactive password
```
Comment (mvdbeek) [inline templates/proftpd.conf.j2]: I don't think you need line 26 to 28, if people should login with their galaxy credentials.
At least that's how we run this on our server.

File: templates/proftpd.conf.j2 [hunk 3]
```diff
@@ -17,6 +17,18 @@ Group                           nogroup
 # Passive port range for the firewall
 PassivePorts                    {{ proftpd_passive_port_low }} {{ proftpd_passive_port_high }}
 
+{% if proftpd_use_sftp_instead %}
+SFTPEngine                      on
+
+# Configure both the RSA and DSA host keys, using the same host key files that OpenSSH uses.
+SFTPHostKey                     /etc/proftpd/ssh_host_keys/rsa
```
Comment (mvdbeek) [inline templates/proftpd.conf.j2]: Can we generate those in a task?

Comment (zfrenchee) [inline templates/proftpd.conf.j2]: yes, I'm doing that locally. Let me add a task. Note: It takes a very long time...

File: templates/proftpd.conf.j2 (hunk 2 above)
Comment (zfrenchee) [inline templates/proftpd.conf.j2]: I believe I do need lines 26/27. Let me get back to you on this.

**Comment by zfrenchee** [comment]
Unresolved: 
- https://github.com/galaxyproject/ansible-galaxy-extras/pull/85#discussion_r72605419
- https://github.com/galaxyproject/ansible-galaxy-extras/pull/85#discussion_r72603951

**Comment by bgruening** [comment]
I marked it as WIP for the time being. @zfrenchee let me know if you need help here, would be nice to get this into the next stable release of the Docker image.

**Comment by zfrenchee** [comment]
Does someone know ansible well enough to be able to answer this: 
https://github.com/ansible/ansible/pull/16920

**Comment by zfrenchee** [comment]
otherwise I think this one is ready for another look through from the maintainers @mvdbeek @bgruening @jmchilton

File: tasks/proftpd.yml [hunk 4]
```diff
@@ -15,3 +15,33 @@
...
+- name: Load Proftpd SFTP modules
+  lineinfile: dest=/etc/proftpd/modules.conf line="{{ item }}"
+  with_items:
+    - LoadModule mod_sftp.c
+    - LoadModule mod_sftp_pam.c
+    - LoadModule mod_sftp_sql.c
+  when: proftpd_use_sftp
+
+- name: Generate new SSH keys (rsa)
+  shell: ssh-keygen -b 2048 -t rsa -f /ssh_host_keys/rsa -N ""
```
Comment (bgruening) [inline tasks/proftpd.yml]: Do we need to install `openssh-client` with some task?

Comment (zfrenchee) [inline tasks/proftpd.yml]: I think it ships with ubuntu by default?

Comment (bgruening) [inline tasks/proftpd.yml]: Can be - but this role should be useful for other as well. Was just an idea.

File: templates/proftpd.conf.j2 (hunk 2 above)
Comment (zfrenchee) [inline templates/proftpd.conf.j2]: Removed.

File: tasks/proftpd.yml (hunk 1 above)
Comment (zfrenchee) [inline tasks/proftpd.yml]: done.

File: defaults/main.yml [hunk 5]
```diff
@@ -148,6 +148,7 @@ proftpd_files_dir: /export/galaxy-central/database/ftp
 proftpd_ftp_port: 21
 proftpd_passive_port_low: 30000
 proftpd_passive_port_high: 40000
+proftpd_use_sftp_instead: false
```
Comment (zfrenchee) [inline defaults/main.yml]: done.

**Comment by bgruening** [comment]
@mvdbeek is this ready to go?

**Comment by zfrenchee** [comment]
@bgruening @mvdbeek @martenson @jmchilton I believe this is good to go, but please take another look through. I'd love to merge this as soon as possible.

**Comment by jmchilton** [comment]
Cool beans - thanks @zfrenchee !

**Comment by mvdbeek** [comment]
@bgruening @zfrenchee sorry, was in vacation. Very nice, thanks a lot!

**Comment by zfrenchee** [comment]
@jmchilton, 
@bgruening and I have been speaking via gitter, and have decided that instead of replacing ftp with sftp, it would be ideal to have them both. Keep an eye out for another PR in the near future moving over to that strategy. If it's alright, I think we should turn them both on by default, instead of having a `use_sftp` flag.

**Comment by mvdbeek** [comment]
@zfrenchee does it not work now? Seems everything is there ?!

**Comment by mvdbeek** [comment]
I see, there's an example [here](https://serverfault.com/questions/570961/proftpd-configuration-to-run-ftp-ftes-ftps-and-sftp-at-the-same-time-on-differen)
Yeah, that would be cool 👍

**Pull request merged**